pkg/resource/deploy/providers/registry.go

```diff
@@ -266,7 +266,18 @@ func (r *Registry) Diff(urn resource.URN, id resource.ID, olds, news resource.Pr
 
 	// Create a reference using the URN and the unknown ID and fetch the provider.
 	provider, ok := r.GetProvider(mustNewReference(urn, UnknownID))
-	contract.Assertf(ok, "'Check' must be called before 'Diff'")
+	if !ok {
+		// In this case, we are doing a diff as part of evaluating the fanout of a delete-before-replace operation.
```

Comment:
What is "this case"?  How did we know that from `!ok`?

Reply:
Correct me if I'm wrong, Pat, this is partly to check my own understanding...

When evaluating the fanout of a DBR operation, we must walk the condemned resource's dependencies in reverse, before the engine has seen them. There's nothing new here, we did this previously for DBR. Pat's twist is that we're now invoking Diff on these resources as we walk them.

Suppose we had a resource graph that looks like this:

```
    A
    |
    B
    |
    C
```

Where `B` is a provider resource. If `A` gets DBR'd, the engine is going to call `Diff` on `B` to understand if a change to `A` is going to have any effect. However, since we are operating in reverse, the engine has not seen `B` yet and, as such, has not had the opportunity to call `Check` on it. `Check` on a provider is what actually loads the provider. So, `r.GetProvider` will only succeed on providers that the engine has seen and called `Check` on.

In this case, `!ok` means "the engine has not yet called `Check` on this provider". Its presence in the checkpoint inductively implies that this reference isn't totally bogus, so it's safe to assume that this provider does exist, just later in the plan.

The remedy here is to just use the old provider for the given provider resource. The provider registry loads all providers on startup, so that provider is guaranteed to be there.

Reply:
> In this case, !ok means "the engine has not yet called Check on this provider". Its presence in the checkpoint inductively implies that this reference isn't totally bogus, so it's safe to assume that this provider does exist, just later in the plan.
> 
> The remedy here is to just use the old provider for the given provider resource. The provider registry loads all providers on startup, so that provider is guaranteed to be there.

This is correct.